src/v/datalake/coordinator/coordinator.cc

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/coordinator/coordinator.h"
+
+#include "base/vlog.h"
+#include "container/fragmented_vector.h"
+#include "datalake/coordinator/state_update.h"
+#include "datalake/logger.h"
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+#include "storage/record_batch_builder.h"
+
+namespace datalake::coordinator {
+
+namespace {
+coordinator::errc convert_stm_errc(coordinator_stm::errc e) {
+    switch (e) {
+    case coordinator_stm::errc::not_leader:
+        return coordinator::errc::not_leader;
+    case coordinator_stm::errc::shutting_down:
+        return coordinator::errc::shutting_down;
+    case coordinator_stm::errc::apply_error:
+        return coordinator::errc::stm_apply_error;
+    case coordinator_stm::errc::replication_error:
+        return coordinator::errc::timedout;
+    }
+}
+} // namespace
+
+std::ostream& operator<<(std::ostream& o, coordinator::errc e) {
+    switch (e) {
+    case coordinator::errc::not_leader:
+        return o << "coordinator::errc::not_leader";
+    case coordinator::errc::shutting_down:
+        return o << "coordinator::errc::shutting_down";
+    case coordinator::errc::stm_apply_error:
+        return o << "coordinator::errc::stm_apply_error";
+    case coordinator::errc::timedout:
+        return o << "coordinator::errc::timedout";
+    }
+}
+
+ss::future<> coordinator::stop_and_wait() {
+    as_.request_abort();
+    return gate_.close();
+}
+
+checked<ss::gate::holder, coordinator::errc> coordinator::maybe_gate() {
+    ss::gate::holder h;
+    if (as_.abort_requested() || gate_.is_closed()) {
+        return errc::shutting_down;
+    }
+    return gate_.hold();
+}
+
+ss::future<checked<std::nullopt_t, coordinator::errc>>
+coordinator::sync_add_files(
+  model::topic_partition tp, chunked_vector<translated_offset_range> entries) {
+    if (entries.empty()) {
+        vlog(datalake_log.debug, "Empty entry requested {}", tp);
+        co_return std::nullopt;
+    }
+    auto gate = maybe_gate();
+    if (gate.has_error()) {
+        co_return gate.error();
+    }
+    vlog(
+      datalake_log.debug,
+      "Sync add files requested {}: [{}, {}], {} files",
+      tp,
+      entries.begin()->start_offset,
+      entries.back().last_offset,
+      entries.size());
+    auto sync_res = co_await stm_.sync(30s);
+    if (sync_res.has_error()) {
+        co_return convert_stm_errc(sync_res.error());
+    }
+    auto added_last_offset = entries.back().last_offset;
+    auto update_res = add_files_update::build(
```

Comment:
how do we think about the concurrency here, are we fine with making a false positive decision about the update that can not be applied ? (i guess we are fine as the translator will retry ?) Maybe it is worth a comment ?

Reply:
Yea, if there is a race, it will fail to apply to the STM at apply time. The apply will effectively do a CAS-like check to make sure the entries can be added. At that point this error is a signal for the worker to resync the last added offset

Reply:
Done